File: solutions/python/perfect-numbers/1/perfect_numbers.py
def classify(number: int) -> str:
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """
    if number < 1:
        raise ValueError(
            "Classification is only possible for positive integers.")

    aliquot_sum = sum(i for i in range(1, number) if number % i == 0)

    return {
        aliquot_sum == number: 'perfect',
        aliquot_sum > number: 'abundant',
        aliquot_sum < number: 'deficient'
    }[True]

File: solutions/ruby/perfect-numbers/1/perfect_numbers.rb
module PerfectNumber
  def self.classification(number, aliquot_sum)
    {
      number == aliquot_sum => 'perfect',
      number > aliquot_sum => 'deficient',
      number < aliquot_sum => 'abundant'
    }[true]
  end


  def self.factors(number)
    (1...number).reduce(Set.new) { |acc, n|
      if (number % n).zero?
        acc << n
      end
      acc
    }.to_a
  end

  def self.classify(number)
    raise RuntimeError.new() if number <= 0

    classification(number, factors(number).sum)
  end
end